packages/server/src/services/uniswap-v3/fetcher.ts

```diff
@@ -0,0 +1,262 @@
+        throw new Error('No pools returned.');
+      }
+    } catch (error) {
+      throw makeSdkError(`Could not fetch top performing pools.`, error);
+    }
+  }
+
+  async getPoolDailyData(
+    poolId: string,
+    start: Date,
+    end: Date,
+  ): Promise<GetPoolDailyDataResult> {
+    const startDate = toDateInt(start);
+    const endDate = toDateInt(end);
+    
+    try {
+      const { poolDayDatas } = await this.sdk.getPoolDailyData({
+        id: poolId,
+        orderBy: 'date',
+        orderDirection: 'asc',
+        startDate,
+        endDate,
+      });
+
+      if (poolDayDatas == null) {
+        throw new Error('No pools returned.')
+      }
+
+      return poolDayDatas;
+    } catch (error) {
+      throw makeSdkError(`Could not fetch daily data for pool ${poolId}.`, error);
+    }
+  }
+
+  async getPoolHourlyData(
+    poolId: string,
+    start: Date,
+    end: Date,
+  ): Promise<GetPoolHourlyDataResult> {
+    const startTime = toDateInt(start) - 1;
+    const endTime = toDateInt(end);
+
+    try {
+      const { poolHourDatas } = await this.sdk.getPoolHourlyData({
+        id: poolId,
+        orderBy: 'periodStartUnix',
+        orderDirection: 'asc',
+        startTime,
+        endTime,
+      });
+
+      if (poolHourDatas == null) {
+        throw new Error('No pools returned.')
+      }
+
+      return poolHourDatas;
+    } catch (error) {
+      throw makeSdkError(`Could not fetch hourly data for pool ${poolId}.`, error);
+    }
+  }
+
+  async getHistoricalDailyData(
+    poolId: string,
+    start: Date,
+    end: Date = new Date(),
+  ): Promise<GetPoolDailyDataResult> {
+    const startData = await this.getPoolDailyData(poolId, start, end);
+    if (startData.length === 0) {
+      throw new HTTPError(
+        404,
+        `Could not fetch any historical data for the given timeframe. Make sure the window is at least 1 day.`
+      );
+    }
+
+    let dailyData = startData; // accumulator
+    let cursorEndDate = dailyData[dailyData.length - 1]?.date; // date of last data in current page
+    let lastStartDate = start; // start date from previous page
+    const endDateTimestamp = Math.floor(end.getTime() / 1000); // final date we want data for
+
+    // Keep fetching until we pass the end date
+    while (
+      cursorEndDate &&
+      cursorEndDate <= endDateTimestamp &&
+      Math.floor(lastStartDate.getTime() / 1000) <= endDateTimestamp
+    ) {
+      // cache old length to compare later
+      const oldLength = dailyData.length;
+
+      // skip ahead 24 hours, cache for next query
+      lastStartDate = new Date(cursorEndDate * 1000 + dayMs);
+      const moreDailyData = await this.getPoolDailyData(poolId, lastStartDate, end);
+      dailyData = [...dailyData, ...moreDailyData];
+      cursorEndDate = dailyData[dailyData.length - 1]?.date; // set new cursor
+
+      // Nothing more to add
+      // @kkennis whats the scenario where there would be no more data to fetch?
```

Comment:
Scenario here is that we either hit the end of the window or there is no more data (we are in the present day). Kind of being defensive here b/c of some weird intraday overlaps with the timestamps. The `UniswapDailyData` timestamp e.g. is not guaranteed to be at 00:00 UTC time, but we should always round to 00:00 UTC time ourselves.

Reply:
We definitely should be able to precalculate windows, we just need to make sure to handle the bounds properly, and re-zip the results in the correct order